### ISSUE TITLE
Optimize TypeScript generated enums

### ISSUE THREAD

**Issue by bhollis** [terser/terser#1064]
**Bug report or Feature request?**

This is a feature request to find a way to recognize the structure of generated code from TypeScript to enable subsequent constant folding optimizations. I'm not super familiar with how to add rules to Terser, but in my head a path forward would be to somehow recognize that the IIFE pattern generated by TypeScript initializes an object's properties, and that information could then be consumed by the other constant folding passes.

Note that I think TypeScript generates the IIFE so that the enum name can be mangled within it, reducing the code size when the enum structure is preserved. It just so happens to defeat constant folding optimizations.

**TypeScript source**

```
enum FooEnum {
    ONE,
    TWO,
    THREE
}

console.log(FooEnum.ONE, FooEnum[0])
```

**`terser` input / generated TypeScript output**

```
"use strict";
var FooEnum;
(function (FooEnum) {
    FooEnum[FooEnum["ONE"] = 0] = "ONE";
    FooEnum[FooEnum["TWO"] = 1] = "TWO";
    FooEnum[FooEnum["THREE"] = 2] = "THREE";
})(FooEnum || (FooEnum = {}));
console.log(FooEnum.ONE, FooEnum[0]);
```

**`terser` output or error**

```
var E;!function(E){E[E.ONE=0]="ONE",E[E.TWO=1]="TWO",E[E.THREE=2]="THREE"}(E||(E={})),console.log(E.ONE,E[0]);
```

**Expected result**

```
console.log(0,"ONE");
```

**Comment by jridgewell** [comment]
This would requires a more robust eval implementation to understand what's going on inside the closure. We can actually already inline if the TS output weren't so convoluted.

**Comment by bhollis** [comment]
That's right, and another avenue is to switch what TypeScript emits, but ideally it'd be both inline-able and minifyable in place if it can't be inlined.

**Comment by bo4arovpavel1989** [comment]
You could use enum with `const` keyword, if you didn't access enum member using a string literal? That would shorten ts output

**Comment by bhollis** [comment]
That's right, and that helps. But you lose the ability to look up enum names, and in general `const enum` is poorly supported in TS.
